src/Perlang.Tests.Integration/Typing/TypingTests.cs

```diff
@@ -124,6 +124,72 @@ public void var_declaration_with_initializer_correctly_infers_type_from_assignme
             Assert.Equal("System.Int32", result);
         }
 
+        [Fact]
+        public void var_declaration_supports_reassignment_after_variable_is_defined()
+        {
+            string source = @"
+                var i = 123;
+                i = 456;
+
+                print i;
+            ";
+
+            string result = EvalReturningOutputString(source);
+            Assert.Equal("456", result);
+        }
+
+        [Fact]
+        public void var_declaration_detects_reassignment_of_incoercible_type()
+        {
+            string source = @"
+                var i = 123;
+                i = ""foo"";
+
+                print i;
+            ";
+
+            var result = EvalWithValidationErrorCatch(source);
+            var exception = result.Errors.FirstOrDefault();
+
+            Assert.Single(result.Errors);
+            Assert.Matches("Cannot assign System.String to variable defined as 'System.Int32'", exception.Message);
+        }
+
+        [Fact]
+        public void var_declaration_supports_reassignment_to_nil_for_reference_type()
+        {
+            // TODO: Think about whether these are the semantics we want or whether we should treat reference types as
+            // non-nullable right from the start. We can always support a Nullable<String> or something if we want to
+            // support it _for explicit cases_ where it might be needed.
+            string source = @"
+                var s = ""foo"";
+                s = nil;
+
+                print s;
+            ";
+
+            var output = EvalReturningOutputString(source);
+
+            Assert.Equal("nil", output);
+        }
```

Comment:
> Other use cases are more tricky. For example, should we or should we not allow `nil` assignment by default for reference types? We'll try to decide on this before merging this PR.

In other words, this. :point_up: We need to figure out whether this is what we want or not. Having it like this is the easiest way out, but it should actually be quite simple to forbid all `nil` assignments also if we like. We could actually give it a try and see how many tests it breaks. :bulb:

It's probably much easier to forbid `nil` assignment at this early stage in the Perlang development than trying to get rid of it after-the-fact, which experience from other languages has proven to be quite hard (not to say _impossible_). (Well, C# has made it possible to default to non-nullability by default which is quite an achievement.)

Reply:
The following is now in place:

- Reassignments of variables to `nil` emits a warning.
- Using `nil` as a variable initializer likewise emits a warning
- Bug fix: Using `nil` for local variable inference (`var s = nil`) is now properly detected as an invalid condition, causing a compilation error. There is simply _no way_ for the compiler to determine what type the user was intending to use in that case, so it's better to just fail fast.
- Passing `nil` as a function parameter also emits a warning.

The PR got a bit more complex because of this. I'll try to extract the infrastructure for emitting warnings into a separate PR in a day or two, since it's an isolated piece of art that doesn't necessarily need to be glued together with the rest like this.

One big caveat at the moment is also the following semantic detail:

- In REPL mode, warnings are treated as warnings and just produce some noise; the code will still run as if the warning wasn't there.
- In script mode, *warnings are treated as errors* (imagine `-Werror` being enabled with `gcc`) and _there is no way to turn it off_. The latter might be a bit unpleasant and we could try to fix a `-Wno-error`-flag or something when we have the infra for warning extracted to an isolated PR.

Reply:
> One big caveat at the moment is also the following semantic detail:
>
> - In REPL mode, warnings are treated as warnings and just produce some noise; the code will still run as if the warning wasn't there.
> - In script mode, *warnings are treated as errors* (imagine `-Werror` being enabled with `gcc`) and _there is no way to turn it off_. The latter might be a bit unpleasant and we could try to fix a `-Wno-error`-flag or something when we have the infra for warning extracted to an isolated PR.

This was now fixed. I added a `-Wno-error=nil-usage` flag that you can use to demote usage of `nil` to just a warning instead of an error.

The REPL still always treats it as a warning which makes sense in my world view of how a REPL should "feel" to the user.

One caveat still: the `EvalHelper` methods will consider all warnings _warnings_ by default, not errors. We could change this (and perhaps we should) but it does berak about 5 unit tests. Some of them could easily be fixed but not others. We would probably have to figure out a way to disable/enable warnings for individual tests to accomplish that (perhaps by moving them to be `Program` integration tests and passing in the proper `-Wno-error` flag). But, that's a different story for another day.

> I'll try to extract the infrastructure for emitting warnings into a separate PR in a day or two, since it's an isolated piece of art that doesn't necessarily need to be glued together with the rest like this.

This part I'll try to deal with before we get this merged though.

Reply:
> This part I'll try to deal with before we get this merged though.

For reference, infrastructure was extracted to #190 and #191, both of which are merged by now.